src/emmy/complex/impl.cljc

```diff
@@ -0,0 +1,616 @@
+#_"SPDX-License-Identifier: GPL-3.0"
+
+;; /**
+;;  * @license Complex.js v2.1.1 12/05/2020
+;;  *
+;;  * Copyright (c) 2020, Robert Eisele (robert@xarg.org)
+;;  * Dual licensed under the MIT or GPL Version 2 licenses.
+;;  **/
+
+(ns emmy.complex.impl
+  "This namespace provides the primitive implementation of complex
+   arithmetic for Emmy. The authors are indebted to Robert Eisele's
+   `Complex.js` implementation from which we have borrowed freely and
+   which is licensed as follows:
+
+   * Copyright (c) 2020, Robert Eisele (robert@xarg.org)
+   * Dual licensed under the MIT or GPL Version 2 licenses."
+  (:refer-clojure :exclude [abs zero?])
+  (:require [emmy.generic :as g]
+            [emmy.util :as u]
+            [emmy.value :as v]))
+
+(declare equal?)
+
+(deftype Complex [re im]
+  v/IKind
+  (kind [_] :emmy.complex/complex)
+
+  v/Numerical
+  (numerical? [_] true)
+
+  #?@(:clj [Object
+            (equals [a b] (equal? a b))]
+      :cljs [IEquiv
+             (-equiv [a b] (equal? a b))
+
+             IPrintWithWriter
+             (-pr-writer
+              [z writer _]
+              (write-all
+               writer
+               "#emmy/complex "
+               (str [(.-re z) (.-im z)])))]))
+
+(def ZERO (Complex. 0 0))
+(def ONE (Complex. 1 0))
+(def I (Complex. 0 1))
+(def -I (Complex. 0 -1))
+(def INFINITY (Complex. ##Inf ##Inf))
+(def NAN (Complex. ##NaN ##NaN))
+(def LN2 (Math/log 2))
+
+(def ^:private PI:2 (/ Math/PI 2))
+(def ^:private PI:4 (/ Math/PI 4))
+
+(defn equal?
+  "Test for complex equality with another object."
+  [^Complex a b]
+  (cond (instance? Complex b)
+        (and (v/= (.-re a) (.-re b))
+             (v/= (.-im a) (.-im b)))
+
+        (v/real? b)
+        (and (g/zero? (.-im a))
+             (v/= (.-re a) b))
+
+        :else (v/= a b)))
```

Comment:
will this ever bottom out at `equals` and recurse back to `equal?`, causing a loop?

Reply:
Good question. I needed (I think) a way to allow the specializations for (e.g.) letting a complex number `=` a quaternion where the 1,i parts match and the j,k parts are zero. I'll look in to this

Reply:
I have imitated the implementation given for quaternions now, so there's no scary recursion possible.

Reply:
This turned out to be a rabbit hole. A quaternion knows how to tell whether it is equal to a complex number, and it installs a method for v/= for (quaternion, complex) as well as (complex, quaternion). This is fine. 

There's a test in quaterion_test:
```clojure
(is (= i**i
               (g/expt q/I q/I)
               (g/expt q/J q/J)
               (g/expt q/K q/K))
            "i^i matches quaternion implementation for each of the components.")
```
wherein we call ordinary `=` on one complex and 3 quaternion values. How does it work in Clojure, since this is `clojure.core/=`? We see [here](https://github.com/clojure/clojure/blob/08a2d9bdd013143a87e50fa82e9740e2ab4ee2c2/src/jvm/clojure/lang/Util.java#L31) that if either of the operands are `IPersistentCollection`s, we go [here](https://github.com/clojure/clojure/blob/08a2d9bdd013143a87e50fa82e9740e2ab4ee2c2/src/jvm/clojure/lang/Util.java#L122), which solves the riddle: Quaternions are persistent collections, and while Complex numbers are not, the equiv will use the Quaternion implementation to finish the job. This doesn't work in ClojureScript, so we can solve this problem in one of a few ways: (1) Make the unit test use v/= (my preference), (2) Add the original fall-through logic you have just questioned above (the _status quo ante_), (3) equip Complex with enough behaviors to get this to work the way it does in Clojure (overkill IMO since the swapped comparison behavior in the persistent collection case I consider surprising). 

For the present, I'm going to make it so that the CLJS implementation does what it used to do, to keep the PR single-issue.